### PR TITLE
refactor go routine for project region output (#1479)

### PR DESCRIPTION
**Pull request by jsilverio22**
### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [x] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/acorn/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/acorn/blob/main/CONTRIBUTING.md#commits)
- [ ] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [ ] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)
#1479

### CODE REVIEW

File: pkg/project/operations.go [hunk 1]
```diff
@@ -190,37 +191,50 @@ func List(ctx context.Context, opts Options) (projects []string, warnings map[st
 	return projects, warnings, nil
 }
 
-func GetDetails(ctx context.Context, opts Options, projectNames []string) (result []DetailProject, err error) {
-	entryCh := make(chan DetailProject)
-	errCh := make(chan error)
+func GetDetails(ctx context.Context, opts Options, projectNames []string) (projects []DetailProject, warnings map[string]error, err error) {
```
Comment (ibuildthecloud) [inline pkg/project/operations.go:194]: Why have the warnings return when you are just coping the `err` field to the map?  Can you just make the `err` public `Err` and the client read that? I'm not sure what the extra data structure and logic buys you.

Comment (jsilverio22) [reply pkg/project/operations.go]: that makes sense, removed the warnings map

Is the purpose of the warnings return in the List function that there are 2 sources they can be coming from local or hub servers and it serves to differentiate where they are coming from?

File: pkg/project/operations.go [hunk 2]
```diff
@@ -190,37 +191,45 @@ func List(ctx context.Context, opts Options) (projects []string, warnings map[st
...
-			opts.Project = proj
+		wg.Add(1)
+		go func(projectName string, opts Options) {
+			defer wg.Done()
+			// Launch a goroutine for each project to retrieve its information
+			opts.Project = projectName
 			c, err := Client(ctx, opts)
 			if err != nil {
-				errCh <- err
+				logrus.Debugf("unable to get client for %s: %s", projectName, err)
```
Comment (njhale) [inline pkg/project/operations.go:223]: nit: should this be a warn level log? or possibly let this bubble up in a `DetailProject.Err` with a `nil` `Project` field?

File: pkg/cli/project.go [hunk 3]
```diff
@@ -88,6 +88,13 @@ func (a *Project) Run(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		return err
 	}
+
+	for _, project := range projectDetails {
+		if project.Err != nil {
+			logrus.Warnf("Could not list projects from [%s]: %v", project.FullName, project.Err)
```
Comment (njhale) [inline pkg/cli/project.go:94]:
```suggestion
			logrus.Warnf("Could not list details of project [%s]: %v", project.FullName, project.Err)
```

File: pkg/cli/project.go [hunk 4]
```diff
@@ -88,6 +88,13 @@ func (a *Project) Run(cmd *cobra.Command, args []string) error {
 	if err != nil {
 		return err
 	}
+
+	for _, project := range projectDetails {
```
Comment (njhale) [inline pkg/cli/project.go:92]: nit: I would do this after sorting so we get deterministic-ish output.

File: pkg/cli/project.go [hunk 5]
```diff
@@ -88,6 +88,13 @@ func (a *Project) Run(cmd *cobra.Command, args []string) error {
...
+		if project.Err != nil {
```
Comment (njhale) [inline pkg/cli/project.go:93]: If we failed to get project details, is it possible that `project.Project == nil`? If so, it looks like the for-loop below can still hit a nil pointer:

```go
for _, projectItem := range projectDetails {
		supportedRegions := projectItem.Project.GetSupportedRegions()
		defaultRegion := projectItem.Project.GetRegion()
		for i, supportedRegion := range supportedRegions {
			if supportedRegion == defaultRegion {
				supportedRegions[i] = supportedRegion + "*"
			}
		}
		out.Write(projectEntry{
			Name:    projectItem.FullName,
			Default: defaultProject == projectItem.FullName,
			Regions: supportedRegions,
		})
}
```

maybe we should add some checks before trying to access it there?

**Review by njhale — CHANGES_REQUESTED** [review]
Nice! Thanks for putting this together

Just a few small questions/comments before I approve.

File: pkg/project/operations.go (hunk 2 above)
Comment (jsilverio22) [reply pkg/project/operations.go]: yeah I was trying to follow the pattern of the List func where it just debug logs when local kubeconfig client is invalid but I think it makes more sense to warn here instead in this case.

File: pkg/cli/project.go (hunk 5 above)
Comment (jsilverio22) [reply pkg/cli/project.go]: good point, I consolidated the err check and now check for nil projectItem in this loop

File: pkg/project/operations.go [hunk 6]
```diff
@@ -190,37 +191,45 @@ func List(ctx context.Context, opts Options) (projects []string, warnings map[st
 	return projects, warnings, nil
 }
 
-func GetDetails(ctx context.Context, opts Options, projectNames []string) (result []DetailProject, err error) {
-	entryCh := make(chan DetailProject)
-	errCh := make(chan error)
+func GetDetails(ctx context.Context, opts Options, projectNames []string) (projects []DetailProject, err error) {
```
Comment (njhale) [inline pkg/project/operations.go:194]: Ah, forgot to post this comment:

This function always returns a `nil` error, I think it's safe to drop the `error` return param.

**Review by njhale — APPROVED** [review]
LGTM! Thanks for the quick turnaround!

**Pull request merged**